### PR TITLE
UW-319

### PR DESCRIPTION
**Pull request by maddenp-noaa**
This PR provides the following general updates to the codebase and tooling:

- The repo content is now organized to be built as a conda package, and steps are outlined in `README.md` to perform the build, create an environment, etc.
- `README.md` also outlines steps for creating a _development shell_ in which all dependency packages (e.g. `f90nml`, `jinja2`) required by the code are installed in a special conda environment and code formatters and quality tools can be run on the `uwtools` codebase. This provides a consistent mechanism for all developers to obtain a conda environment to develop in, and to run all tools with identical settings.
- Settings for code quality tools are now consolidated in the `src/pyproject.toml` file. [pyproject.toml](https://peps.python.org/pep-0518/) is an emerging standard for configuring Python build- and test-tools.
- Installed and configured code-formatting tools are `black`, `isort`, and `docformatter`; code-quality tools are `pylint`, `mypy`, and `pytest`.
- [`black`](https://pypi.org/project/black/) provides consistent formatting, automatically applying whitespace, commas, parentheses, line wraps, etc. [`isort`](https://pycqa.github.io/isort/) organizes `import` statements, creating separate standard-library / 3rd-party / local package sections, and sorting entries (in a way acceptable to `pylint`, by the way). [`docformatter`](https://github.com/PyCQA/docformatter) formats docstring that follow the [PEP 257](https://peps.python.org/pep-0257/) standard, with Sphinx-style field lists (e.g. descriptions of function arguments). Line-length or all tools is set to 100 columns. The long-term utility of `docformatter` for `uwtools` will depend on decisions we make on standardizing our docstring, and we may eventually opt to remove it. For now, it was useful to format many of our existing docstrings. As outlined in `README.md`, running `make format` in a development shell will format all of the `uwtools` code with these three tools.
- [`pylint`](https://github.com/pylint-dev/pylint) was already used on `uwtools`. Configuration now resides in `src/pyproject.toml`, and `make lint` will run `pylint` independent of the other code-quality tools. Many unnecessary `pylint` directives have been removed from the codebase.
- [`mypy`](https://github.com/python/mypy) is a static typechecker tool for Python, available via the `make typecheck` command in a development shell. Type-annotating code can be a _progressive_ exercise: As annotations are added, `mypy` has more information to work with and can point out more possible issues. I have started adding annotations, but there's plenty still to do.
- [`pytest`](https://docs.pytest.org/en/7.4.x/) was also already used on `uwtools`. Configuration now resides in `src/pyproject.toml`, and `make unittest` will run `pytest` independent of the other code-quality tools. Some unit tests were updated to use some `pytest` features like [fixtures](https://docs.pytest.org/en/6.2.x/fixture.html) (both user-defined, and built-in fixtures like [`tmp_path`](https://docs.pytest.org/en/7.1.x/how-to/tmp_path.html), [`capsys`](https://docs.pytest.org/en/7.1.x/how-to/capture-stdout-stderr.html), [`caplog`](https://docs.pytest.org/en/7.1.x/how-to/logging.html)) and [parameterization)[https://docs.pytest.org/en/7.3.x/how-to/parametrize.html]. Tests now reside in `src/uwtools/tests`, and there's a 1:1 correspondence between `uwtools` main modules and corresponding test modules. Code coverage is 100%.
- The `uwtools` version number is now maintained in conda's `recipe/meta.yaml` file, and documentation-build logic (via both GitHub Actions and Read the Docs) are updated to obtain the version number from this single canonical value.
- Because the `uwtools` conda-package build logic is configured to run all code-quality tests as a prerequisite to a successful package build, GitHub actions have been updated to run a `make package` command to build `uwtools` with Python 3.9, 3.10, and 3.11 and run all the tests with each. Developers can also run `make package` locally (not in a development shell, but with conda activated -- see `README.md`) to test all supported Python versions.
- Because `black`, `isort`, and `docformatter` have been run across the entire codebase, there are **many** diffs related strictly to formatting. Also, I have lightly reorganized some modules so that member functions, methods, etc. are lexicographically ordered (aka in alphabetical order, aka sorted), rather than randomly (or chronologically) positioned in the source file.
- When `uwtools` is installed, either via conda or pip, `setuptools` is called upon (configuration is in `src/setup.py`) to create "entry-point console scripts", aka CLI tools. These are currently named `atparse-to-jinja2`, `experiment-manager`, `run-forecast`, `set-config`, `template`, and `validate-config`. They should be called directly, not prefixed with `python` or `python3`. We have a Jira ticket to consolidate these into a single modal CLI.
- Where modules contained both CLI-handling code (`argparse`, etc.) and business logic, these have been split: `src/uwtools/cli` contains the CLI-handling modules, and business logic resides elsewhere under `src/uwtools`.
- For (hopefully temporary) backward compatibility, the `scripts/` directory currently contains minimal shims for invoking code in the legacy way, e.g. `python3 scripts/set_config.py`.
- CLI-tool switches are updated to use `--kebab-case` switches instead of `--snake_case`. The former is nearly universally preferred by tools in the wild.

If you are interested in reviewing specific diffs, you might consider using GitHub's option to suppress whitespace diffs.

### CODE REVIEW

File: src/uwtools/tests/cli/test_templater.py [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+    for line in expected.split("\n"):
+        assert line_in_lines(line, actual)
+
+
+@pytest.mark.parametrize(
+    "sw", [ns(d="-d", i="-i", v="-v"), ns(d="--dry-run", i="--input-template", v="--verbose")]
+)
+def test_set_template_verbosity(sw, capsys):
+    infile = fixture_path("nml.IN")
+    # #PM# WHAT TO DO ABOUT /dev/null BELOW?
```
Comment (maddenp-noaa) [inline src/uwtools/tests/cli/test_templater.py:177]: This relates to the default log file value being supplied in the `parse_args()` function -- I've set that to `/dev/null` to avoid having the logger create files in source-code directories, but we need to decide what to do about this in general.

File: environment.yaml
[the remark this answers is not in the document]
Comment (christinaholtNOAA) [reply environment.yaml:1]: Let's open a Jira Ticket?

File: recipe/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    - pyyaml 6.0.*
+    - {{ python }}
+test:
+  requires:
+    - black
+    - coverage
+    - isort
+    - mypy
+    - pylint
+    - pytest
```
[the remark this answers is not in the document]
Comment (christinaholtNOAA) [reply recipe/meta.yaml:30]: One reason we've pinned them in the past was because we wanted consistency with supported packages on RDHPCS platforms, i.e., pylint passes on Hera/Jet, but not on GitHub actions because of the version difference.

We may need to go on deep dive (along with some of our EPIC colleagues) to see what makes sense for supporting on all the platforms.

File: recipe/.gitignore
[the remark this answers is not in the document]
Comment (christinaholtNOAA) [reply recipe/.gitignore:1]: I'm seeing a bunch of these types of comments here in the PR, which are super useful. I'm curious if this type of commentary can go into a header of the files, or perhaps even a `recipe/README.md` that points users to the `condev` documentation on what each file is responsible for?

Apologies if I missed it, but it might also be nice to boldly state somewhere that this is a `condev` enabled repo and what that means and what advantages it provides users and developers. Again, just linking to the `condev` docs for that would also likely suffice. 

In other places, your PR comment mentions that a file is a standard "whatever" kind of file for use with tool x. Would it be appropriate to drop that into the file header for helpful hints to newcomers?

File: src/pyproject.toml
[the remark this answers is not in the document]
Comment (christinaholtNOAA) [reply src/pyproject.toml:1]: This is one of the ones I was referring to in a previous comment about describing how files in our repo work with other standard tools. This file is new to me, and I likely wouldn't have thought to google it, considering that it might just be another file for this project.

File: src/uwtools/apps/srw.py [hunk 3]
```diff
@@ -1,76 +1,74 @@
-#!/usr/bin/env python3
-#pylint: disable=unused-import, unused-variable, unused-argument, useless-parent-delegation
-# remove these disables once implemented
-'''
-This file contains the specific drivers for a particular app, using the facade pattern base class.
-'''
+"""
+This file contains the specific drivers for a particular app, using the facade
+pattern base class.
+"""
```
[the remark this answers is not in the document]
Comment (christinaholtNOAA) [reply src/uwtools/apps/srw.py:4]: Just to be clear for my understanding, `black` doesn't change the _length_ of comment lines? I'd assume it wouldn't change the content of any line it made a change to the length of.

Do you have suggestions on how we might all stick to this same convention if we can't lean on `black` to do it for us (assuming that's what you meant)? For example, I don't think my `.vimrc` is going to let me change handle 100 columns for code differently than 80 columns for comments.

File: src/uwtools/cli/experiment_manager.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        help="Path to YAML, bash/ini, or namelist configuration file",
+        metavar="FILE",
+        type=cli_helpers.path_if_file_exists,
+    )
+    optional = parser.add_argument_group("optional arguments")
+    optional.add_argument(
+        "-l",
+        "--log-file",
+        # #PM# WHAT TO DO ABOUT THIS LOGFILE PATH?
+        default="/dev/null",  # os.path.join(os.path.dirname(__file__), "forecast.log"),
```
[the remark this answers is not in the document]
Comment (christinaholtNOAA) [reply src/uwtools/cli/experiment_manager.py:56]: Yes. Let's please continue this overall discussion.

File: src/uwtools/config.py [hunk 5]
```diff
@@ -235,29 +229,26 @@ def dereference(self, ref_dict=None, full_dict=None):
...
 
-                                filter: {repr(e).split()[-1][:-3]}
-                                key: {key}
+filter: {repr(e).split()[-1][:-3]}
+key: {key}
 
-                                Define the filter before proceeding.
-                                """)
+Define the filter before proceeding.
+""".strip()
```
[the remark this answers is not in the document]
Comment (christinaholtNOAA) [reply src/uwtools/config.py:213]: This is such a trivial thing, I know, but I really hate as a developer seeing blocks of beautiful Python code broken up by unindented multiline strings that don't follow proper indentation. This is totally a product of my own baggage, but is there any chance we could leave this with the `dedent`?

File: src/uwtools/config.py [hunk 6]
```diff
@@ -298,12 +289,16 @@ def dereference_all(self):
...
-        ''' Check if the string contains a float, int, boolean, or just
+    def str_to_type(self, str_: str) -> Union[bool, float, int, str]:
+        """Check if the string contains a float, int, boolean, or just
         regular string. This will be used to automatically convert
         environment variables to data types that are more convenient to
-        work with.'''
+        work with."""
 
+        # #PM# CONSIDER MANDATING YAML-COMPATIBLE VALUES (e.g. https://yaml.org/type/bool.html)
+        #      AND VALDATING THEM WITH THE YAML PARSER?
```
[the remark this answers is not in the document]
Comment (christinaholtNOAA) [reply src/uwtools/config.py:342]: Yep. I'm cool with that.

File: src/uwtools/config.py [hunk 7]
```diff
@@ -342,178 +335,99 @@ def update_values(self, new_dict, dict_to_update=None):
...
-        Recursively parse which keys in the object are complete (set_var), which keys have 
-        unfilled jinja templates (jinja2_var), and which keys are set to empty (empty_var). 
-        '''
-
-        if not isinstance(config_dict, dict):
-            return
+    def iterate_values(
+        self,
+        config_dict: dict,
+        set_var: List[str],
```
[the remark this answers is not in the document]
Comment (christinaholtNOAA) [reply src/uwtools/config.py]: I'm historically horrible at naming things! These sorts of suggestions are 💯 always welcome!

File: src/uwtools/config.py [hunk 8]
```diff
@@ -342,178 +335,99 @@ def update_values(self, new_dict, dict_to_update=None):
...
             self.update(self._load())
             self.parse_include()
 
-    def _load(self, config_path=None):
-        ''' Load the user-provided Fortran namelist path into a dict
-        object. '''
+    def _load(self, config_path: Optional[str] = None) -> dict:
+        """Load the user-provided Fortran namelist path into a dict
+        object."""
+        # #PM# THIS SEEMS TO SILENTLY FAIL IF A NAMELIST IN THE FILE IS INVALID.
```
[the remark this answers is not in the document]
Comment (christinaholtNOAA) [reply src/uwtools/config.py:399]: You are correct.

File: src/uwtools/config.py [hunk 9]
```diff
@@ -342,178 +335,99 @@ def update_values(self, new_dict, dict_to_update=None):
...
-            cfg = f90nml.read(file_name).todict(complex_tuple=False)
-
-        cfg = dict(cfg)
-        self.from_ordereddict(cfg)
-        return cfg
+        # #PM# SINCE config_path in ctor IS ALSO OPTIONAL, THIS ASSERTION IS REQUIRED (PROBABLY BAD)
+        assert config_path is not None
+        with open(config_path, "r", encoding="utf-8") as f:
+            cfg = f90nml.read(f).todict(complex_tuple=False)
+        # #PM# WHY NOT JUST KEEP OrderedDict OBJECT?
```
[the remark this answers is not in the document]
Comment (christinaholtNOAA) [reply src/uwtools/config.py:405]: This was a requirement in the older versions we were testing. I believe we were seeing failures with Python 3.7 in our matrix of versions in GitHub Actions.

File: src/uwtools/config.py [hunk 10]
```diff
@@ -582,31 +580,133 @@ def _format_output(self):
...
+            if (outfile_type == "INI" and input_depth > 2) or (
+                outfile_type == "F90" and input_depth != 2
+            ):
+                log.critical(err_msg)
+                raise ValueError(err_msg)
+
+            # #PM# HOW IS THIS POSSIBLE?
+        # if input_depth > output_depth:
+        #     log.critical(f"{user_args.outfile} not compatible with {user_args.input_base_file}")
+        #     raise ValueError("Set config failure: output object not compatible with input file")
```
[the remark this answers is not in the document]
Comment (christinaholtNOAA) [reply src/uwtools/config.py:724]: Nice call. I think we're checking the more important part above in that the maximum allowable input depths for a given output language are not exceeded.

File: src/uwtools/drivers/forecast.py [hunk 11]
```diff
@@ -57,30 +51,32 @@ def create_model_config(self, config_obj, outconfig_file, base_file=None):
...
-                        values override any settings in base file
-            outconfig_file: location of output configuration file
-            base_file: optional path to file to use as a base file
-        '''
+            config_file: Path to user configuration file
+            outconfig_file: Path to output configuration file
+            base_file: Path to optional file with base configuration
+        """
+
+        # #PM# DEAL WITH ISSUE OF config_file / config_obj TYPE.
```
[the remark this answers is not in the document]
Comment (christinaholtNOAA) [reply src/uwtools/drivers/forecast.py:60]: I think it's fine for now. We do have work on the backlog to make sure that this one comes together in a working fashion. It would be good to have a review of what "drivers" are meant to accomplish before, or as part of, that work. We can decide what we need (config obj or config file) at which points in the process of the driver.

File: src/uwtools/files/interface/file_manager.py [hunk 12]
```diff
@@ -1,25 +1,23 @@
...
-        """copies source to destination"""
+    @abstractmethod  # pragma: no cover
+    def copy(self, source: List[File], destination: List):
+        """Copies source to destination."""
         raise NotImplementedError
 
-    @classmethod
-    def get_file_manager(cls, _type: Prefixes):
-        """returns a file manager with source and destination"""
+    @staticmethod
```
[the remark this answers is not in the document]
Comment (christinaholtNOAA) [reply src/uwtools/files/interface/file_manager.py:18]: I have been trying to understand the use cases between these two. I understand what `@staticmethod` does I think. I don't think I've fully grasped what a good use case for `@classmethod` is just yet. 

I suppose here you clarify that `@classmethod` might be able to access class attributes, if need be. We don't (at this point, I _think_) have any class attribute use cases in our code base to make this necessary. They kind of give me the same heebie-jeebies as global variables and I try to stay away from them when I can. Is that the wrong response?

File: src/uwtools/tests/cli/test_atparse_to_jinja2.py [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+from pytest import raises
+
+from uwtools.cli import atparse_to_jinja2
+from uwtools.tests.support import fixture_path
+
+# NB: Ensure that at least one test exercises both short and long forms of each
+#     CLI switch.
+
+
+@pytest.mark.parametrize("sw", [ns(i="-i"), ns(i="--input-template")])
```
[the remark this answers is not in the document]
Comment (christinaholtNOAA) [reply src/uwtools/tests/cli/test_atparse_to_jinja2.py:19]: I know this wasn't the takeaway of this particular comment, but TIL about the few diffs between `SimpleNamespace` instead of `argparse.Namespace`. (https://github.com/python/cpython/issues/83257)

File: src/uwtools/utils/file_helpers.py [hunk 14]
```diff
@@ -1,54 +1,37 @@
...
+        raise RuntimeError(msg) from e
+
+    # Try to rename existing run directory if option is rename.
 
-    # Try to rename existing run directory if caller chooses rename
     try:
         if exist_act == "rename" and os.path.isdir(run_directory):
-            now = datetime.now()
-            save_dir = run_directory + now.strftime("_%Y%m%d_%H%M%S")
+            now = dt.now()  # #PM# SHOULD WE UTCNOW?
```
[the remark this answers is not in the document]
Comment (christinaholtNOAA) [reply src/uwtools/utils/file_helpers.py:31]: I think in this case, it might be helpful to users to have this particular thing adopt the time zone used by the platform where it's being run. It is only meant to be a unique string added to the path when we move it out of the way.

File: src/uwtools/tests/test_config.py [hunk 15]
```diff
@@ -0,0 +1,861 @@
...
+    """Compare two config objects."""
+    ext = ".%s" % ("nml" if fmt == "F90" else fmt).lower()
+    cfgobj = help_cfgclass(ext)(fixture_path(f"simple{ext}"), log_name="test_compare_config")
+    cfgobj.log.addHandler(logging.StreamHandler(sys.stdout))
+    cfgobj.log.setLevel(logging.INFO)
+    cfgobj.compare_config(cfgobj, salad_base)
+    if fmt in ["F90", "YAML"]:
+        assert msg_in_caplog("salad:        how_many:  - 12 + 12", caplog.records)
+    else:
+        assert not caplog.records  # #PM# BUT WHY NOT IN INI?
```
[the remark this answers is not in the document]
Comment (christinaholtNOAA) [reply src/uwtools/tests/test_config.py:141]: I have none. It's not ringing any bells.

**Comment by christinaholtNOAA** [comment]
I've really only read through the comments you left in the "Conversation" tab so far. I still haven't looked in depth at the code. What I'm seeing is pretty great, and will guide us in a variety of Q&A/Discussion sessions to have as a team moving forward from this refactor.

It's a nice time to do all this IMO. We have a couple of tools that do what we need them to (in general), and now we have a proper code base. We can see how these concepts are applied with tangible examples that some of us have become more familiar with.

File: src/uwtools/apps/srw.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (christinaholtNOAA) [reply src/uwtools/apps/srw.py:4]: It only just occurred to me that the `.vimrc` file could be set to 80 columns, if a user so desired. Then `black` could fix all the code lines to 100 columns.

[the remark this answers is not in the document]
Comment (maddenp-noaa) [reply src/uwtools/apps/srw.py:4]: Is that a good or a bad thing? It's potentially kind of annoying for the vim user but IMO good for the team in that everyone can expect consistent formatting when the code reaches them. I also think that vim and other editors can be configured to honor per-project settings, that e.g. a vim user could have 80 columns as their general default, but would get 100 when working on `uwtools`. I'm not familiar with the `.editorconfig` system, but I suspect that vim should be able to honor that, and I've updated the one in this repo to agree with the `black` settings.

File: src/uwtools/config.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (maddenp-noaa) [reply src/uwtools/config.py:399]: On further reflection, it's not clear to me that the "invalid" namelist file is actually invalid, given how tolerant the namelist definition is of strange use patterns (and the characters that appear between namelists in a file containing namelists is really wild-west territory where pretty much anything goes. For example, I just learned that string values in namelists do not necessarily need to be quoted, which is just awful (but typical Fortran). So, what I thought was invalid might be valid after all.

File: src/uwtools/config.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (maddenp-noaa) [reply src/uwtools/config.py:342]: Opened https://jira.epic.oarcloud.noaa.gov/browse/UW-362 for this.

File: src/uwtools/config.py [hunk 16]
```diff
@@ -298,12 +289,16 @@ def dereference_all(self):
...
+        """Check if the string contains a float, int, boolean, or just
         regular string. This will be used to automatically convert
         environment variables to data types that are more convenient to
-        work with.'''
+        work with."""
 
+        # #PM# CONSIDER MANDATING YAML-COMPATIBLE VALUES (e.g. https://yaml.org/type/bool.html)
+        #      AND VALDATING THEM WITH THE YAML PARSER?
+        # #PM# str_to_type() IS AN INACCURATE NAME: THE VALUES RETURNED ARE OBJECTS, NOT TYPES.
+        #      reify_str() MIGHT BE A GOOD NAME.
```
[the remark this answers is not in the document]
Comment (maddenp-noaa) [reply src/uwtools/config.py:344]: Also a subject of newly-opened https://jira.epic.oarcloud.noaa.gov/browse/UW-362

File: src/uwtools/config.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (maddenp-noaa) [reply src/uwtools/config.py:405]: @christinaholtNOAA {{OrderedDict}} dates from Python 3.2, so I'd be curious (but may never know) what was failing under 3.7.

In any case, and to ask a fundamental question: Python 3.8 is the earliest version that is not end-of-life. What's our stance on supporting EOL Pythons? IMO we should _not_ because they are not even getting security updates, and we should not encourage continued use of them. Should we have a policy of only tracking/testing currently-supported versions? If so, and assuming we can test successfully against 3.8-3.11, should we consider removing this {{OrderedDict}} -> {{dict}} translation? I could create a Jira ticket to explore/implement that...

File: src/uwtools/config.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (maddenp-noaa) [reply src/uwtools/config.py:724]: @christinaholtNOAA Yeah, I'm going to remove this code and resolve the conversation. I'll be happy to address anything that comes up, but I don't think it will.

File: src/uwtools/tests/cli/test_templater.py (hunk 1 above)
Comment (maddenp-noaa) [reply src/uwtools/tests/cli/test_templater.py:177]: I'll just leave it for now. It will disappear if/when we revamp logging.

File: src/uwtools/tests/test_config.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (maddenp-noaa) [reply src/uwtools/tests/test_config.py:141]: @christinaholtNOAA I found it: For F90 and YAML, the `"how_many": "12"` `str` value in the `assets` test (inline `pytest`) fixture is a mismatch with the `12` `int` value in the (external file) fixture, but for INI -- which I assume parses everything a `str` objects? -- it's a match. So I'm updating the tests that use this fixture so that we don't have this weird edge case.

File: src/uwtools/tests/test_config_validator.py [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+        config_validator.config_is_valid(
+            config_file=fixture_path(f"schema_test_{cfgtype}.yaml"),
+            validation_schema=fixture_path("schema/workflow.jsonschema"),
+            log=Logger(),
+        )
+        is boolval
+    )
+
+
+# #PM# EVERYTHING BELOW TESTS jsonschema, NOT OUR CODE. DO WE NEED IT?
```
[the remark this answers is not in the document]
Comment (maddenp-noaa) [reply src/uwtools/tests/test_config_validator.py:37]: @christinaholtNOAA Per our conversation this afternoon, and given that this only tests `jsonschema` and not our code, I'm removing this bit of testing.

File: src/uwtools/files/interface/file_manager.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (maddenp-noaa) [reply src/uwtools/files/interface/file_manager.py:22]: @christinaholtNOAA class and static methods are both legit and shouldn't cause you any distress. They're (very) good when used properly.

Also, FWIW, and possibly slightly pedantic: There's not really such a thing as a "global variable" in Python. In the most general sense, they are "module-level variables" and, since Python modules are just (singleton) _objects_, they're really just instance variables on module objects, and you shouldn't hate them. For example, the `uwtools.utils.memory` module has a `MAP` module-level variable, and introspecting the state of affairs:
```
>>> from uwtools.utils import memory
>>> type(memory)
<class 'module'>
>>> isinstance(memory, object)
True
>>> memory.MAP
{'KB': 1000, 'MB': 1000000, 'GB': 1000000000}
```
We see that the `memory` object that we've imported is an `object` of more-specific type `module`, and it has an instance variable called `MAP`. Nothing scary here.

File: src/uwtools/apps/srw.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (maddenp-noaa) [reply src/uwtools/apps/srw.py:3]: @christinaholtNOAA I've now included `docformatter` as part of the `make test` target, configured to match the `black` 100-column line length. I think the results look right. The only leaves `#`-prefixed comments unaddressed, so we'll need to wrap those manually unless we find another way to handle them. On the plus side, `pylint` should complain about > 100 column `#` comments, so I think the only unaddressed case will be _unnecessarily short_ `#` comments, which we can re-flow to 100 columns manually where necessary, for now.

File: src/setup.py [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+
+setup(
+    entry_points={
+        "console_scripts": [
+            "atparse-to-jinja2 = %s.cli.atparse_to_jinja2:main" % name_py,
+            "experiment-manager = %s.cli.experiment_manager:main" % name_py,
+            "run-forecast = %s.cli.run_forecast:main" % name_py,
+            "set-config = %s.cli.set_config:main" % name_py,
+            "template = %s.cli.templater:main" % name_py,
+            "validate-config = %s.cli.validate_config:main" % name_py,
```
[the remark this answers is not in the document]
Comment (WeirAE) [reply src/setup.py:35]: Very useful, I hadn't actually implemented this before and had been looking, so thanks!

**Review by WeirAE — APPROVED** [review]
Reiterating my out-of-date review's awe at this refactoring. LGTM.

**Review by christinaholtNOAA — APPROVED** [review]
LGTM. Thanks @maddenp-noaa!
